Use a single `g3w-client/gulpfile.js` to compile all client plugins

volterra79 opened the pull request:
## List of changes

- **new function** `browserify_plugin(pluginName)`,
   closes: https://github.com/g3w-suite/g3w-client/issues/484 

- **refactor function** `g3wsdk.info()`,
    closes: https://github.com/g3w-suite/g3w-client/issues/442

- **update docs** `README.md`,
    closes: https://github.com/g3w-suite/g3w-client/issues/112

- **deprecate npm task** `npm run build:plugins` → replaced by: `npm run run build`

- **deprecate npm task** `npm run watch:plugins` → replaced by: `npm run run dev`

volterra79 committed:
:sparkles: Introduce watch:new-plugins and strem tasks to handle plugins wath chnges and build

volterra79 committed:
:fire: Remove all reference to gulpfile.js plugin and add new utility function browserify_plugin.

volterra79 committed:
:sparkles: Try to found a way to check if config change and are add or remove plugin to work with in deveopment

volterra79 committed:
:sparkles: Build plugin.js in production mode  (minified) when run build:plugins task

volterra79 committed:
:sparkles: add new gulp task add:build_dev_plugins to add dinamically plugin to develop without restart dev scritp on client. Isolate also select_plugin prompt method utility

volterra79 committed:
Use Reflect on Proxy trap

volterra79 commented:
@Raruto we can think also how add plugins version on g3wsdk.info

Raruto committed:
Merge branch 'dev' into issue_484

Raruto committed:
multiple outputs (dev mode): `outputFolder` + `pluginsFolder`

Raruto committed:
remove task: `build:dev_plugins`

Raruto committed:
wrong `overridesFolder`

Raruto committed:
log info

Raruto committed:
comments

volterra79 committed:
:bulb: Comments and clean code

Raruto committed:
new gulp task: `check:node_modules`

Raruto committed:
comments

Raruto committed:
comments

Raruto committed:
log

Raruto committed:
spacing

Raruto committed:
comments

Raruto committed:
emit `_version.js` file into client plugins

Raruto committed:
draft variable: `g3wsdk.loaded_plugins`

Raruto committed:
indentation

Raruto committed:
restore browsersync reload

Raruto committed:
update `g3wsdk.info`

Raruto committed:
comments

Raruto committed:
Create RELEASE.md

Raruto committed:
Update RELEASE.md

Raruto committed:
move publishing info into `README.md`

Raruto committed:
subsection

Raruto committed:
links

Raruto committed:
add since tags

Raruto committed:
get plugin version from `PluginRegistry`

Raruto committed:
spacing

volterra79 committed:
deprecate method

Raruto committed:
set version `onbefore` registering plugin

Raruto committed:
Merge branch 'issue_484' of https://github.com/g3w-suite/g3w-client into issue_484

volterra79 committed:
space

Raruto committed:
replace `vm.$watch` with ternary condition

volterra79 committed:
:bulb: Comments

Raruto committed:
Delete applicationservice.js

Raruto committed:
Merge branch 'issue_484' of https://github.com/g3w-suite/g3w-client into issue_484

volterra79 committed:
:bulb: Comments and clean code

volterra79 committed:
Clean code

volterra79 committed:
Clean code

Raruto committed:
jsdocs

volterra79 committed:
Add new core plugin qprocessing and write version.js on plugins folder

volterra79 committed:
Renamed version.js file on src/plugins folder to _version.js

volterra79 committed:
Emite complte event when all plugins are ready

Raruto commented on gulpfile.js:
```diff
@@ -75,6 +75,7 @@ const default_plugins = [
   'editing',
   'openrouteservice',
   'qplotly',
+  'qprocessing', //@since 3.9.0
```
@volterra79 this is not a core plugin 👉 [g3w-admin-processing](https://github.com/g3w-suite/g3w-admin-processing)

volterra79 replied:
@Raruto you are right. Please remove it

Raruto committed:
prevent adding a new watcher

Raruto committed:
watch `src/plugins/_version.js`

Raruto committed:
await task completion: `'deploy-plugins'`

Raruto committed:
function closure

Raruto committed:
remove gulp task: `'add:dev_plugins'`

Raruto committed:
reduce delta

Raruto committed:
reduce delta

Raruto committed:
reduce delta

Raruto committed:
prevent fatal error (no selected plugins)

Raruto committed:
remove npm tasks: `"build:plugins"`, "watch:plugins

Raruto reviewed (approved):
@volterra79 for me it's quite ok.

For any further improvements, consider opening a new PR..

volterra79 commented:
@Raruto plugins version are undefined

```js
...
- plugins:
    - editing: __undefined__
    - qtimeseries: __undefined__
    - qprocessing: __undefined__
....
```

Raruto committed:
backcomp

Raruto committed:
hardcoded path for: `/src/plugins/_version.js`

Raruto committed:
Merge branch 'dev' into issue_484

Raruto commented:
@volterra79 please do final checks and merge this..

(so we start testing, and make future adjustments if necessary)

volterra79 commented:
> @volterra79 please do final checks and merge this..
> 
> (so we start testing, and make future adjustments if necessary)

@Raruto we can convert it to Draft

Raruto commented:
I don't think we can amended to a later release.

This PR is necessary to solve the [`regeneratorRuntime`](https://github.com/g3w-suite/g3w-client/pull/365) issue / dependency on client plugins.

Raruto committed:
Merge branch 'dev' into issue_484

Raruto committed:
Merge branch 'dev' into issue_484

volterra79 committed:
Merge branch 'dev' into issue_484

# Conflicts:
#	gulpfile.js

volterra79 committed:
Merge branch 'issue_484' of https://github.com/g3w-suite/g3w-client into issue_484

# Conflicts:
#	gulpfile.js

volterra79 committed:
Change comment

volterra79 committed:
Clean code

volterra79 committed:
Fix otuputFolder template string value

volterra79 committed:
Clean code

volterra79 committed:
Using insertGlobals and insertGlobalVars browserify options to insert global varible for version (plugins). Reomve creation of _version.js on each plugins

volterra79 committed:
Remove creation of _version.js in each plugins

volterra79 committed:
Use a version.js file to show g3w-client version. Gulfile include branch and commti.
For p

volterra79 committed:
Version file with last commit and branch info

volterra79 committed:
Get version and set plugin version to config from version.js file create by gulpfile

volterra79 committed:
Add create_version common to app and plugins

volterra79 committed:
show only first seven characters of commit has

volterra79 committed:
fix require cache fro package.json version

volterra79 committed:
Comments and set version in diffente way in case of new release

volterra79 committed:
Set branch and commit hash

volterra79 committed:
Clean code and comments

volterra79 committed:
Introduced insertGlobalVars with __git__ global variable to insert as add info to version

volterra79 committed:
Clean code

volterra79 committed:
Test commit to check if rebuild get new commit hash on client info

volterra79 committed:
Clean code and set opts object

volterra79 committed:
Nee to set get git info sync not async

volterra79 committed:
Add pre-commit task to listen commit event on git

volterra79 committed:
Revert "Add pre-commit task to listen commit event on git"

This reverts commit c8f2f7aab9c77ec1a4c8385a4f02dfef0226f7d8.

volterra79 committed:
Load all plugin installed locally on src/plugins

volterra79 committed:
Merge branch 'dev' into issue_484

# Conflicts:
#	src/store/plugins.js

volterra79 committed:
Fix merge

volterra79 committed:
Set Warning message aligned with v3.9.1 gulpfile message

volterra79 committed:
Set Warning message aligned with v3.9.1 gulpfile message

volterra79 committed:
Add info and git methods to plugin

volterra79 committed:
extract getInfo() from plugin

volterra79 committed:
Align version forma as g3w-admin

volterra79 committed:
Clean code

volterra79 committed:
Merge branch 'dev' into issue_484

Raruto committed:
Merge branch 'dev' into issue_484

Raruto committed:
Merge branch 'dev' into issue_484

Raruto committed:
code dedupe

Raruto committed:
remove `event-stream`

Raruto committed:
revert 063298b581d8f3729668233bac4ef1b6fa06dae5

Raruto committed:
update warning

Raruto committed:
slim down promise chain

Raruto committed:
make use of a single `browserify_plugin`

Raruto committed:
fix `runSequence` + restore `src/plugins/_version.js`

Raruto committed:
YAGNI

Raruto commented on gulpfile.js:
```diff
@@ -82,145 +83,124 @@ const dev_plugins = Array.from(
-        }
-        return null;
-      },
-    },
```
@volterra79 I removed the `insertGlobalVars` solution because:

1. it generates a lot of duplicate and unintelligible code (ref: `commonjs` code within `app.min.js` and `plugin.js`)
2. it is significantly slower than `gulp-replace`
3. the overall build process was no longer synchronous (ref: `build:client` task + `runSequence` and any nested `await`/`async` function)

Raruto committed:
show current branch while developing

Raruto committed:
show current client branch while developing

Raruto committed:
move `gulp-sourcemaps` after `gulp-merge`

Raruto committed:
wrong merge src

Raruto committed:
refactor function `PluginsRegistry~_setup(name, pluginConfig)`

- convert to plain js promises
- remove jquery dependency
- make it debugger friendly (prevent `eval()` usage)

Raruto committed:
move src

Raruto committed:
refactor function: refactor function `PluginsRegistry~init(options)`

- reduce nesting level

Raruto committed:
missing id

Raruto committed:
fix invalid type

Raruto committed:
Avoid mutating prop directly

volterra79 committed:
Clean code

volterra79 committed:
Clean code
